File: Data_Structure_java/LinkedList/CircularlyLinkedList.java
public class CircularlyLinkedList<E> {
	private Node <E> tail = null;
	private int size = 0;
	public CircularlyLinkedList() {}

	//access method
	public int size() {return size;}
	public boolean isEmpty() {return size == 0;}
	public E first() {
		if (isEmpty()) return null;
		else return tail.getNext().getElement();
	}
	public E last() {
		if (isEmpty()) return null;
		else return tail.getElement();
	}

	public void rotate(){
		if (tail != null) tail = tail.getNext();
	}
	public void addFirst(E e){
		if (size == 0) {
			tail = new Node<>(e,null);
			tail.setNext(tail);
		}
		else{
			Node<E> newest = new Node<>(e,tail.getNext());
			tail.setNext(newest);
		}
		size ++;
	}
	public void addLast(E e){
		addFirst(e);
		tail = tail.getNext();
	}
	public E removeFirst(){
		if (isEmpty()) return null;
		Node<E> head = tail.getNext();
		if (head == tail) tail = null;
		else tail.setNext(head.getNext());
		size --;
		return head.getElement();
	}

	public void printlist(){
		Node <E> temp = tail;
		System.out.print("Head: ");
		while (temp.getNext() != tail){
			temp = temp.getNext();
			System.out.print(temp.getElement()+" ");
		}
		System.out.print("Tail: "+temp.getNext().getElement());
		System.out.println();
	}
}

File: Data_Structure_python/LinkedList/CircularlyLinkedList.py
class CircularlyLinkedList(Node):
	def __init__(self):
		self.tail = None
		self.size = 0

	def isEmpty(self): return self.size == 0

	#method
	def rotate(self):
		if self.tail != None:
			self.tail = self.tail.getNext()  

	def addFirst(self,data):
		if (self.isEmpty()):
			self.tail = Node(data,null)
			self.tail.setNext(self.tail)
		else:
			new = Node(data,self.tail.getNext())
			self.tail.setNext(new)

	def addLast(self,data):
		self.addFirst(data)
		self.tail = self.tail.getNext()

	def removeFirst(self):
		head = self.tail.getNext()
		if (self.isEmpty()):
			self.tail = None
		else:
			self.tail.setNext(head.getNext())
		size -= 1
		return head.getElement()

	def printlist(self):
		temp = self.tail.getNext()
		print("head: ")
		while temp != self.tail:
			print(temp.getElement())